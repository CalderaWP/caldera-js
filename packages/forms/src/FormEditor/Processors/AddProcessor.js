import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Row, fieldAreaFactory} from "@calderawp/components";

export const AddProcessor = ({
								 setNewProcessorType,
								 processorTypes,
								 onCreate,
								 children,
								 value
							 }) => {
	const options = [
		{
			value: null,
			label: '--'
		}
	];

	processorTypes.map(processorType => {
		options.push({
			value: processorType.type,
			label: processorType.type
		});
	});


	const processorTypesField = {
		fieldType: 'select',
		label: 'Processor Type',
		fieldId: 'newProcessorType',
		required: true,
		options,
		onChange: setNewProcessorType,
		value,
	};


	return (
		<Fragment>
			{fieldAreaFactory(processorTypesField, setNewProcessorType)}
			<button onClick={onCreate}
					disabled={!value}
			>
				{children}
			</button>
		</Fragment>
	)
};

AddProcessor.propTypes = {
	processorTypes: PropTypes.array,
	onCreate: PropTypes.func,
	setNewProcessorType: PropTypes.func,
	value: PropTypes.string
};

AddProcessor.defaultProps = {
	processorTypes: [],
	value: ''
};
