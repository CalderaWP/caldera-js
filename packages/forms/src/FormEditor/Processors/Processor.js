import React, { Fragment, Component } from 'react';
import { HorizontalForm } from '../../HorizontalForm/HorizontalForm';
import { TabPanel } from '@wordpress/components';
import { Row, Column,FormFieldsAutoComplete } from"@calderawp/components";
import PropTypes from 'prop-types';

export class Processor extends Component {
	state = {
		activeTab: 'settings'
	};

	onSetTab = activeTab => {
		this.setState({ activeTab });
	};

	tabs = [
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-processor-settings-tab-btn'
		},
		{
			name: 'conditionals',
			title: 'Conditionals',
			className:
				'caldera-processor-conditionals-tab-btn'
		}
	];

	/**
	 * Prepare props for the processor editing form
	 *
	 * @return {Processor.props}
	 */
	formProps = () => {
		let {props} = this;
		const {fields,form} = this.props;
		const formFields = form.fields;
		fields.forEach( (field,i) => {
			const {fieldType} = field;
			if( FormFieldsAutoComplete.IDENTIFIER === fieldType ){
				props.fields[i] = {
					...field,
					form
				}
			}
		});
		return props;
	};

	render() {
		const { onClose, onRemove } = this.props;
		return (
			<Fragment>
				<Row>
					<TabPanel
						className="caldera-processor"
						activeClass="active-tab"
						onSelect={this.onSetTab}
						initialTabName={'settings'}
						tabs={this.tabs}
					>
						{() => {
							const {activeTab} = this.state;
							const tab =  this.tabs.find( t => t.name === activeTab );
							const { name } = tab;
							if ('settings' === name) {
								return (
									<HorizontalForm
										{...this.formProps()}
										className={'caldera-processor-settings'}
									/>
								);
							}
							return (
								<div
									className={'caldera-processor-conditionals'}
								>
									Conditionals
								</div>
							);
						}}
					</TabPanel>
				</Row>
				<Row>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-close'}
							onClick={onClose}>
							Close
						</button>
					</Column>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-remove'}
							onClick={onRemove}>
							Remove
						</button>
					</Column>
				</Row>
			</Fragment>
		);
	}
}

//should have label!
Processor.propTypes = {
	...HorizontalForm.propTypes,
	initialActiveTab: PropTypes.string,
	onRemove: PropTypes.func,
	type: PropTypes.string,
	id: PropTypes.string,
};

Processor.defaultProps = {
	...HorizontalForm.defaultProps,
	initialActiveTab: 'settings'
};
