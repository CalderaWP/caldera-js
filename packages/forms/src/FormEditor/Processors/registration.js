

const defaultRegistry = [];

/**
 * Add a processor type to form editor
 *
 * @param {*} processorType Defintion of processor type
 * @param {Array} registry Optional. Collection to add to -- generally you leave this at default.
 * @return {Array}
 */
export const registerProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		return [ processorType,...registry];
	}

	return registry;

};

/**
 * Get a processor type
 *
 * @param {string} processorType
 * @param {Array} registry
 * @return {*}
 */
export const unregisterProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		processorType = processorType.type;
	}
	const index = registry.findIndex( p => p.type === processorType );
	if( 0<= index ){
		registry.splice(index,1);
		return [ ...registry];
	}
	return registry;
};

/**
 * Get all processors from collection
 *
 * @param registry
 * @return {Array}
 */
export const getProcessors = (registry = defaultRegistry) => {
	return registry;
};
