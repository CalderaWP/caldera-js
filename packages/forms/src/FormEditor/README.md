
## Horizontal Forms
This is a small wrapper around the `<CalderaForm>` component that is supplied an array of fields and each field is shown in its own row.

### Primary Use Case(s)
* A section of Caldera UI that has a horizontal form, form example:
    - Processor UI
    - Field Config UI
* The inspector controls section of a Gutenberg block.


### Props
### Generic Example
```jsx
import {HorizontalForm} from '@calderawp/components';
<HorizontalForm
    fields={[]}
    initialValues={{}}
    onClose={onClose}
    onChange={onChange}
    instanceId={'test-2'}
/>
```

### Components
There are components that wrap `<HorizontalForm />`. They generally add controls for the UI state -- open, close, minimize, expand.

#### `<Processor />`
This is the settings -- to be used in the Caldera Forms editor or similar situation for describe one process of a form.

```js
import {Processor} from '@calderawp/components';

const processor = {
    id: 'p1',
    type: 'mailchimp',
    label: 'Segment One',
    fields: [
    	
    ],
    conditionals: [
    	
    ]
};

<Processor
    fields={processor.fields}
    initialValues={{}}
    onClose={onClose} //called when processor close button is clicked
    onChange={onChange} //called when any field is changed
    initialActiveTab={'conditionals'}//defaults to settings
    instanceId={processor.id} //shou
/>
```
