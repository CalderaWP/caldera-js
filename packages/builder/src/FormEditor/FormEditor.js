import React, {Component, Fragment, createElement} from 'react';
import PropTypes from 'prop-types';
import {Processors, processorsCollectionPropType} from './Processors/Processors';
import {TabPanel} from '@wordpress/components';
import {Row, Column,fieldAreaFactory} from "@calderajs/components";
import {processorTypesPropType} from './propTypes';
import {MainSection} from './MainSection';
import {defaultProcessorTypes} from './Processors/processorTypes/defaultProcessorTypes';
import {FormEntryViewer} from "..";

export class FormEditor extends Component {
	state = {
		activeTab: 'processors',
		newProcessorType: '', //the next processor to be created will use this type
	};

	onSetTab = activeTab => {
		this.setState({activeTab});
	};

	updateProcessors = processors => {
		const {updateForm, form} = this.props;
		updateForm({
			...form,
			processors
		});
	};

	getFormProcessors = () => {
		const {form} = this.props;
		return form.processors ? form.processors : [];
	};

	getFormFields = () => {
		const {form} = this.props;
		return form.fields ? form.fields : [];
	};

	tabs = [
		{
			name: 'editor',
			title: 'Layout',
			className: 'caldera-forms-editor-layout-tab-btn',
			classNameForComponent: 'caldera-forms-editor-layout'
		},
		{
			name: 'entries',
			title: 'Entries',
			className: 'caldera-forms-editor-entries-tab-btn',
			classNameForComponent: 'caldera-forms-editor-entries'
		},
		{
			name: 'processors',
			title: 'Processors',
			className: 'caldera-forms-editor-processors-tab-btn',
			classNameForComponent: 'caldera-forms-editor-processors'

		},
		{
			name: 'mailer',
			title: 'Mail',
			className: 'caldera-forms-editor-mailer-tab-btn',
			classNameForComponent: 'caldera-forms-editor-mailer'
		},
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-forms-editor-settings-tab-btn',
			classNameForComponent: 'caldera-forms-editor-settings'
		},
	];

	getTabs(){
		const {hideTabs} = this.props;
		if( ! hideTabs.length ){
			return this.tabs;
		}

		return  this.tabs.filter( tab =>!hideTabs.includes(tab.name) );
	}


	setNewProcessorType = (newProcessorType) => this.setState({newProcessorType});

	render() {
		const {form, processorTypes, updateForm,entries} = this.props;
		const theProcessorTypes = [
			...processorTypes,
			...defaultProcessorTypes
		];
		return (
			<div>
				<Row>
					<Column>
						<p>{form.name}</p>
					</Column>
				</Row>
				<TabPanel
					className="caldera-processor"
					activeClass="active-tab"
					onSelect={this.onSetTab}
					initialTabName={'processors'}
					tabs={this.getTabs()}
				>
					{tab => {
						const {name, classNameForComponent, title} = tab;
						if ('processors' === name) {
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									<Processors
										processorTypes={theProcessorTypes}
										processors={this.getFormProcessors()}
										form={form}
										formFields={this.getFormFields()}
										updateProcessors={this.updateProcessors}
									/>
								</MainSection>
							);
						}
						if ('settings' === name) {
							const nameField = {
								fieldType: 'text',
								value: form.name,
								label: 'Form Name',
								fieldId: 'formName',
								required: true,
							};
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									{fieldAreaFactory(
										nameField,
										(name) => {
											updateForm({
												...form,
												name
											})
										}
									)}
								</MainSection>
							)
						}
						if ('entries' === name) {
							let entryViewerProps ={
								form,
								noItemsMessage: `No Entries Found For ${form.id}`,
								entries: {}
							};

							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									<FormEntryViewer {...entryViewerProps} />
								</MainSection>
							)
						}
						return (
							<MainSection
								className={classNameForComponent}
								title={title}
							>
								{tab.title}
							</MainSection>
						);
					}}
				</TabPanel>
			</div>
		);
	}
}


FormEditor.propTypes = {
	processorsTypes: processorTypesPropType,
	updateForm: PropTypes.func,
	form: PropTypes.shape({
		id: PropTypes.string,
		name: PropTypes.string,
		fields: PropTypes.array,
		conditionals: PropTypes.array,
		settings: PropTypes.array,
		processors: processorsCollectionPropType,
	}),
	hideTabs: PropTypes.array
};

FormEditor.defaultProps = {
	processorTypes: [],
	updateForm: () => {},
	hideTabs: []
};

