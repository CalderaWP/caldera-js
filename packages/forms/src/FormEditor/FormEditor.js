import React, {Component, Fragment, createElement} from 'react';
import PropTypes from 'prop-types';
import {Processors, processorsCollectionPropType} from './Processors/Processors';
import {TabPanel} from '@wordpress/components';
import {Row, Column,fieldAreaFactory} from "@calderawp/components";
import {processorTypesPropType} from './propTypes';
import {MainSection} from './MainSection';
import {defaultProcessorTypes} from './Processors/processorTypes/defaultProcessorTypes';

export class FormEditor extends Component {
	state = {
		activeTab: 'processors',
		newProcessorType: '', //the next processor to be created will use this type
	};

	onSetTab = activeTab => {
		this.setState({activeTab});
	};

	updateProcessors = processors => {
		const {updateForm, form} = this.props;
		updateForm({
			...form,
			processors
		});
	};

	getFormProcessors = () => {
		const {form} = this.props;
		return form.processors ? form.processors : [];
	};

	getFormFields = () => {
		const {form} = this.props;
		return form.fields ? form.fields : [];
	};

	tabs = [
		{
			name: 'editor',
			title: 'Layout',
			className: 'caldera-forms-editor-layout-tab-btn',
			classNameForComponent: 'caldera-forms-editor-layout'
		},
		{
			name: 'processors',
			title: 'Processors',
			className: 'caldera-forms-editor-processors-tab-btn',
			classNameForComponent: 'caldera-forms-editor-processors'

		},
		{
			name: 'Mailer',
			title: 'mailer',
			className: 'caldera-forms-editor-mailer-tab-btn',
			classNameForComponent: 'caldera-forms-editor-mailer'
		},
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-forms-editor-settings-tab-btn',
			classNameForComponent: 'caldera-forms-editor-settings'

		},
	];


	setNewProcessorType = (newProcessorType) => this.setState({newProcessorType});

	render() {
		const {form, processorTypes, updateForm} = this.props;
		const theProcessorTypes = [
			...processorTypes,
			...defaultProcessorTypes
		];
		return (
			<div>
				<Row>
					<Column>
						<p>{form.name}</p>
					</Column>
				</Row>
				<TabPanel
					className="caldera-processor"
					activeClass="active-tab"
					onSelect={this.onSetTab}
					initialTabName={'processors'}
					tabs={this.tabs}
				>
					{tab => {
						const {name, classNameForComponent, title} = tab;
						if ('processors' === name) {
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									<Processors
										processorTypes={theProcessorTypes}
										processors={this.getFormProcessors()}
										form={form}
										formFields={this.getFormFields()}
										updateProcessors={this.updateProcessors}
									/>
								</MainSection>
							);
						}
						if ('settings' === name) {
							const nameField = {
								fieldType: 'text',
								value: form.name,
								label: 'Form Name',
								fieldId: 'formName',
								required: true,
							};
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									{fieldAreaFactory(
										nameField,
										(name) => {
											updateForm({
												...form,
												name
											})
										}
									)}
								</MainSection>
							)
						}
						return (
							<MainSection
								className={classNameForComponent}
								title={title}
							>
								{tab.title}
							</MainSection>
						);
					}}
				</TabPanel>
			</div>
		);
	}
}


FormEditor.propTypes = {
	processorsTypes: processorTypesPropType,
	updateForm: PropTypes.func,
	form: PropTypes.shape({
		id: PropTypes.string,
		name: PropTypes.string,
		fields: PropTypes.array,
		conditionals: PropTypes.array,
		settings: PropTypes.array,
		processors: processorsCollectionPropType,
	}),
};

FormEditor.defaultProps = {

}

