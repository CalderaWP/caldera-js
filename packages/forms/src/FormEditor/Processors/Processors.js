import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Processor} from "./Processor";
import {Row, fieldAreaFactory} from "@calderawp/components";
import {processorTypesPropType} from './processorTypesPropType';
import processorFactory from './processorTypes/processorFactory';
import {AddProcessor} from './AddProcessor';


export class Processors extends Component {

	state = {
		activeProcessorId: '',
		newProcessorType: ''
	};

	/**
	 * Set which processor is active
	 * @param activeProcessorId
	 */
	setActive = (activeProcessorId) => {
		this.setState({activeProcessorId});
	};

	/**
	 * Check if a processor is active by ID
	 * @param processorId
	 * @return {string|boolean}
	 */
	isActiveProcessor = (processorId) => {
		const {activeProcessorId} = this.state;
		return activeProcessorId && processorId === activeProcessorId;
	};

	/**
	 * When a change in value for a processor is received, send all processors up.
	 *
	 * @param processorId
	 * @param updatedProcessor
	 */
	handleProcessorChange = (processorId, updatedProcessor) => {
		const {processors, updateProcessors} = this.props;
		let processorIndex = processors.findIndex(processor => {
			return processorId === processor.id
		});
		if (-1 !== processorIndex) {
			let processor = processors[processorIndex];
			let update = [
				...processors.splice(processorIndex - 1, 1),
				{
					...processor,
					...updatedProcessor
				}
			];
			updateProcessors(update);
		}


	};

	/**
	 * When a processor is removed, send updated list to change handler.
	 *
	 * @param processorId
	 */
	handleRemoveProcessor = (processorId) => {
		const {processors, updateProcessors} = this.props;
		updateProcessors([...processors.filter(processor => {
			return processorId !== processor.id
		})]);
	};

	/**
	 * When a processor is added, create send update list to change handler

	 */
	handleCreateProcessor = () => {
		const {processors, updateProcessors} = this.props;
		const {activeProcessor,newProcessorType} = this.state;
		const newProcessor = processorFactory(newProcessorType);
		const {id} = newProcessor;
		updateProcessors([...processors, newProcessor]);
		this.setState({activeProcessorId:id,newProcessorType: ''});
	};

	/**
	 * Set the type of the next processor to be created
	 *
	 * @param newProcessorType
	 * @return {*}
	 */
	setNewProcessorType = (newProcessorType) => this.setState({newProcessorType});

	/** @inheritDoc **/
	render() {
		const {activeProcessor,newProcessorType} = this.state;
		const {processors,processorTypes,form} = this.props;
		return (
			<div>
				<div>
					{processors.map(processor => {
						const {
							id,
							fields,
							config,
							type,
							label,
							conditionals
						} = processor;
						if (this.isActiveProcessor(id)) {
							return (
								<Fragment key={id}>
									<Processor
										form={form}
										className={`caldera-forms-active-processor-${id}`}
										fields={fields}
										conditionals={conditionals}
										initialValues={config}
										label={label}
										type={type}
										onChange={(fieldValues) => this.handleProcessorChange(id, fieldValues)}
										onRemove={() => this.handleRemoveProcessor(id)}
										onClose={() => this.setActive('')}
									/>
								</Fragment>
							)
						}
						return (
							<Row key={id}>
								<button
									className={`caldera-forms-choose-processor caldera-forms-choose-processor-${id}`}
									onClick={() => this.setActive(id)}
								>
									{label ? label : type}
								</button>
							</Row>

						)
					})}
				</div>
				<div>
					<AddProcessor
						setNewProcessorType={this.setNewProcessorType}
						processorTypes={processorTypes}
						value={newProcessorType}
						onCreate={this.handleCreateProcessor}
					>
						Add {! newProcessorType ? '' : newProcessorType } Processor
					</AddProcessor>
				</div>
			</div>
		);
	}
}

export const processorsCollectionPropType = PropTypes.arrayOf(
	PropTypes.shape({
		id: PropTypes.string,
		label: PropTypes.string,
		type: PropTypes.string,
		fields: PropTypes.array,
		config: PropTypes.object
	})
);

Processors.propTypes = {
	processorTypes: processorTypesPropType,
	processors: processorsCollectionPropType,
	updateProcessors: PropTypes.func,
	form: PropTypes.object,
	formFields: PropTypes.array
};

Processors.defaultProps = {
	processors: [],
	processorTypes: []
};
