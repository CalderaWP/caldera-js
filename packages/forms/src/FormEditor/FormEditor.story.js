import React, {Component} from 'react';
import { storiesOf } from '@storybook/react';
import { Processor } from './Processors/Processor';
import { Processors } from './Processors/Processors';
import {processorsCollection} from './Processors/processors.fixtures';

import { FormEditor } from './FormEditor';


import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from "@calderawp/components";
storiesOf('FormEditor', module).add('The processor', () => (
	<Processor
		fields={[
			checkboxFieldset,
			selectField,
			checkboxField,
			numberField,
			textField,
			emailField,
			radioField
		]}
		type={'mailchimp'}
		label={'Main Segment'}
		onChange={values => console.log(values)}
		onClose={values => console.log(values)}
		onRemove={values => console.log(values)}
	/>
));

const processorTypes = [
	{
		type: 'apiRequest'
	},
	{
		type: 'redirect'
	}
];

storiesOf('FormEditor', module).add('The processors list', () => (
	<Processors
		processorTypes={processorTypes}
		processors={processorsCollection}
		onChange={values => console.log(values)}
		onClose={values => console.log(values)}
		onRemove={values => console.log(values)}
		formFields={[]}
		form={ {id: 'Name'}}
	/>
));

const form = {
	id: 'a form',
	processors: processorsCollection,
};

class MockFormApp extends Component {
	state = {
		form: {
			id: 'a-form',
			name: 'Form Name',
			processors: processorsCollection,
			fields: [
				checkboxField,
				radioField,
				textField
			]
		}
	};

	updateForm = (form) => this.setState({form});

	render(){
		const {form} = this.state;
		return(
			<FormEditor
				processorTypes={processorTypes}
				updateForm={this.updateForm}
				form={ form }
			/>
		)
	}
}

storiesOf('FormEditor', module).add('The form editor', () => (
	<MockFormApp/>
));


