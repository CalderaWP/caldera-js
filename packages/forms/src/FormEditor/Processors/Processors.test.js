import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Processor} from "./Processor";
import {Processors} from "./Processors";
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';
import {processorsCollection} from './processors.fixtures';

import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from "@calderawp/components";
import {FormEditor} from "../FormEditor";

describe('Processors', () => {
	const processors = [
		{id: 'p1', type: 'Redirect', fields: []},
		{id: 'p2', type: 'Redirect', fields: []},
		{id: 'p3', type: 'Redirect', fields: []},
	];

	const form = {};
	const formFields = [];
	let updateProcessors;
	beforeEach(() => {
		updateProcessors = jest.fn();
	});


	it('Should update active', () => {
		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.instance().setActive('p2');
		expect(component.state('activeProcessorId')).toEqual('p2');
	});

	it('Can remove a processor', () => {
		let values = {};
		const updateProcessors = (updateValues) => {
			values = updateValues;
		};
		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.instance().handleRemoveProcessor('p2');

		expect(values.length).toEqual(2);
	});

	it( 'Can update a processor type', () => {
		const processors = [
			{id: 'p1', type: 'Redirect', fields: []},
			{id: 'p2', type: 'Redirect', fields: [],label: 'fLabel'},
		];
		let values = {};
		const updateProcessors = (updateValues) => {
			values = updateValues;
		};

		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		component.instance().handleProcessorChange('p2', {
			type: 'strident'
		});
		expect(values.find( p => 'p2' === p.id).label).toEqual('fLabel');
		expect(values.find( p => 'p2' === p.id).type).toEqual('strident');
	});

	it( 'Does not call processor update if processor id is invalid', () => {
		const processors = [
			{id: 'p1', type: 'Redirect', fields: []},
			{id: 'p2', type: 'Redirect', fields: [],label: 'fLabel'},
		];

		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		component.instance().handleProcessorChange('p166161111', {
			type: 'strident'
		});
		expect(updateProcessors.mock.calls.length).toEqual(0);
	});

});

describe( 'Processors collection in processors UI', () => {
	const form = {};
	const formFields = [];
	let updateProcessors;
	beforeEach(() => {
		updateProcessors = jest.fn();
	});

	it( 'Renders with a list of processors', () => {
		const component = renderer.create(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('should open processor editor', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		component.find( '.caldera-forms-choose-processor-p2' ).simulate( 'click');
		expect(component.state( 'activeProcessorId') ).toEqual( 'p2');
	});

	it('Calls the processor update callback when a processor changes values', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		const event = {target: {value: 'six'}};

		component.find( '.caldera-forms-choose-processor-p2' ).simulate('click')
		component.find( '.caldera-forms-active-processor-p2 input[type="text"]' ).simulate( 'change',event);
		expect(updateProcessors.mock.calls.length ).toEqual( 1);
	});

	it('Calls the processor update callback when a processor is removed', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		const event = {target: {value: 'six'}};

		component.find( '.caldera-forms-choose-processor-p2' ).simulate('click')
		component.find( '.caldera-processor-remove' ).simulate( 'click' );
		expect(updateProcessors.mock.calls.length ).toEqual( 1);
	});

	it('Closes the processor', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		const event = {target: {value: 'six'}};

		component.find( '.caldera-forms-choose-processor-p2' ).simulate('click')
		component.find( '.caldera-processor-close' ).simulate( 'click' );
		expect(component.state('activeProcessorId') ).toEqual( '');
	});

	it( 'sets newProcessorType', () => {
		const component =
			mount(<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>);
		component.instance().setNewProcessorType('redirect');
		expect( component.state('newProcessorType')).toBe( 'redirect');

	})


	it('Adds new processor when created', () => {
		const component =
			mount(<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>);
		component.setState({newProcessorType: 'autoResponder'});
		component.instance().handleCreateProcessor();
		expect(updateProcessors.mock.calls.length ).toEqual( 1);
		expect(updateProcessors.mock.calls[0][0].find( p => p.type === 'autoResponder').type ).toEqual('autoResponder');


	});



});
