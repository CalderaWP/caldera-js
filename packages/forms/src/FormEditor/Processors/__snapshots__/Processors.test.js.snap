// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processors collection in processors UI Renders with a list of processors 1`] = `
<div>
  <div>
    <div
      className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
      id={undefined}
    >
      <button
        className="caldera-forms-choose-processor caldera-forms-choose-processor-season7Episode2"
        onClick={[Function]}
      >
        Test sending form data to test API
      </button>
    </div>
    <div
      className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
      id={undefined}
    >
      <button
        className="caldera-forms-choose-processor caldera-forms-choose-processor-p2"
        onClick={[Function]}
      >
        Main MailChimp
      </button>
    </div>
  </div>
  <div>
    <div
      className="caldera-field-wrapper caldera-field-wrapper-text"
    >
      <div
        className="components-base-control caldera-field caldera-field-select"
      >
        <div
          className="components-base-control__field"
        >
          <label
            className="components-base-control__label"
            htmlFor="inspector-select-control-4"
          >
            Processor Type
          </label>
          <select
            aria-describedby={undefined}
            className="components-select-control__input"
            id="newProcessorType"
            multiple={false}
            onChange={[Function]}
            value=""
          >
            <option
              value={null}
            >
              --
            </option>
          </select>
        </div>
      </div>
    </div>
    <button
      disabled={true}
      onClick={[Function]}
    >
      Add 
      
       Processor
    </button>
  </div>
</div>
`;
